daemon/oci_linux.go

```diff
@@ -645,7 +645,24 @@ func WithMounts(daemon *Daemon, c *container.Container) coci.SpecOpts {
 			}
 			opts := []string{bindMode}
 			if !m.Writable {
-				opts = append(opts, "ro")
+				rro := true
+				if m.ReadOnlyNonRecursive {
+					rro = false
+					if m.ReadOnlyForceRecursive {
+						return errors.New("mount options conflict: ReadOnlyNonRecursive && ReadOnlyForceRecursive")
+					}
+				}
+				if rroErr := daemon.supportsRecursivelyReadOnly(c.HostConfig.Runtime); rroErr != nil {
+					rro = false
+					if m.ReadOnlyForceRecursive {
+						return rroErr
+					}
+				}
```

Comment:
"Supports RRO" is really a tri-state: yes, no, or unknown. As the runtime features document is optional, and `features.mountOptions == nil` means unknown, it doesn't seem terribly user-friendly to fail without even trying if `m.ReadOnlyForceRecursive`, the kernel supports RRO mounts, and the runtime support is unknown. What if instead we did the following?

| Kernel Supports RRO | Runtime Supports RRO | `ReadOnlyForceRecursive` | Result |
| --- | --- | --- | --- |
| no | - | false | "ro" |
| no | - | true | error |
| yes | no | true | error |
| yes | yes | - | "rro" |
| yes | unknown | false | "ro" |
| yes | unknown | true | "rro" |

That is, try passing the `"rro"` mount option if runtime support is unknown. Best case, the runtime understands the mount option and the mount succeeds. Worst case, the runtime doesn't understand it, passes `"rro"` in the `*data` argument, and the kernel rejects the mount syscall. The end result is the same in the worst case: container start fails with an error.

Reply:
> without even trying if `m.ReadOnlyForceRecursive`

This is because some runtime implementations may just ignore "rro" without raising an error, and such ignorance of `ReadOnlyForceRecursive` may sound like a vulnerability.

In this case, user is suggested to use a runtime that properly supports "rro".